module/zcommon/zfs_fletcher_aarch64_neon.c

```diff
@@ -60,141 +60,185 @@ static void
 {
-	const uint64_t *ip = buf;
-	const uint64_t *ipend = (uint64_t *)((uint8_t *)ip + size);
+	const uint32_t *ip = buf;
+	const uint32_t *ipend = ip + (size / sizeof (uint32_t));
 #if defined(_KERNEL)
-register unsigned char ZERO asm("v0") __attribute__((vector_size(16)));
-register unsigned char ACC0 asm("v1") __attribute__((vector_size(16)));
-register unsigned char ACC1 asm("v2") __attribute__((vector_size(16)));
-register unsigned char ACC2 asm("v3") __attribute__((vector_size(16)));
-register unsigned char ACC3 asm("v4") __attribute__((vector_size(16)));
-register unsigned char TMP1 asm("v5") __attribute__((vector_size(16)));
-register unsigned char TMP2 asm("v6") __attribute__((vector_size(16)));
-register unsigned char SRC asm("v7") __attribute__((vector_size(16)));
+	register unsigned char AA2 asm("v1") __attribute__((vector_size(16)));
+	register unsigned char A3A4 asm("v2") __attribute__((vector_size(16)));
+	register unsigned char BB2 asm("v3") __attribute__((vector_size(16)));
+	register unsigned char B3B4 asm("v4") __attribute__((vector_size(16)));
+	register unsigned char CC2 asm("v5") __attribute__((vector_size(16)));
+	register unsigned char C3C4 asm("v6") __attribute__((vector_size(16)));
+	register unsigned char DD2 asm("v7") __attribute__((vector_size(16)));
+	register unsigned char D3D4 asm("v8") __attribute__((vector_size(16)));
+	register unsigned char SRC asm("v9") __attribute__((vector_size(16)));
+	register unsigned char ZERO asm("v10") __attribute__((vector_size(16)));
+	register unsigned char TMP1 asm("v11") __attribute__((vector_size(16)));
+	register unsigned char TMP2 asm("v12") __attribute__((vector_size(16)));
 #else
-unsigned char ZERO __attribute__((vector_size(16)));
-unsigned char ACC0 __attribute__((vector_size(16)));
-unsigned char ACC1 __attribute__((vector_size(16)));
-unsigned char ACC2 __attribute__((vector_size(16)));
-unsigned char ACC3 __attribute__((vector_size(16)));
-unsigned char TMP1 __attribute__((vector_size(16)));
-unsigned char TMP2 __attribute__((vector_size(16)));
-unsigned char SRC __attribute__((vector_size(16)));
+	unsigned char AA2 __attribute__((vector_size(16)));
+	unsigned char A3A4 __attribute__((vector_size(16)));
+	unsigned char BB2 __attribute__((vector_size(16)));
+	unsigned char B3B4 __attribute__((vector_size(16)));
+	unsigned char CC2 __attribute__((vector_size(16)));
+	unsigned char C3C4 __attribute__((vector_size(16)));
+	unsigned char DD2 __attribute__((vector_size(16)));
+	unsigned char D3D4 __attribute__((vector_size(16)));
+	unsigned char SRC __attribute__((vector_size(16)));
+	unsigned char ZERO __attribute__((vector_size(16)));
+	unsigned char TMP1 __attribute__((vector_size(16)));
+	unsigned char TMP2 __attribute__((vector_size(16)));
 #endif
 
 	kfpu_begin();
 
 	NEON_INIT_LOOP();
 
-	for (; ip < ipend; ip += 2) {
-		NEON_MAIN_LOOP(NEON_DONT_REVERSE);
+	for (; ip < ipend; ip += 4) {
+		NEON_MAIN_LOOP_LOAD(NEON_DO_NOT_REVERSE);
+		PREF1KL1(ip, 1024)
```

Comment:
Have we confirmed that prefetch helps here? It is possible for prefetch instructions to be harmful. In this case, on the last loop iteration, this instruction will have prefetched 1008 bytes into L1 cache that are unlikely to ever be used, which means that we are filling the L1 cache with garbage that we are unlikely to use. This might not show a performance issue in the benchmark, but it could slow down the system elsewhere.

In addition, Linux mainline has had various less than flattering things to say about explicit prefetch instructions over the years:

https://lwn.net/Articles/404103/
https://lwn.net/Articles/444344/

Reply:
@ryao Thanks for pointing out here, I do have some test on the following platform and get performance increase on Kunpeng 920 VM. For Ampere Altra and Aws Gravtion 3, the pre-fetch L1 does not get any improvement. So I agree with you we can remove the prefetch code here.

```
Ampere Altra:
With L1 Prefetch
0 0 0x01 -1 0 20253112240840 20293116392520
implementation   native         byteswap
scalar           4465636046     3986526251
superscalar      6288311844     5471735330
superscalar4     7267898111     5856494177
aarch64_neon     9549582587     7824610103
fastest          aarch64_neon   aarch64_neon

Without
0 0 0x01 -1 0 20361322942120 20362454024320
implementation   native         byteswap
scalar           4441989324     4047774560
superscalar      6273451195     5428114404
superscalar4     7273948180     5848654377
aarch64_neon     9448333032     7895309088
fastest          aarch64_neon   aarch64_neon


Kunpeng 920 VM
With Pre-L1
0 0 0x01 -1 0 28699859899380 28700387328790
implementation   native         byteswap
scalar           3505946470     3087293256
superscalar      4915305602     4253383496
superscalar4     5473091929     4556896233
aarch64_neon     5199690074     4318504180
fastest          superscalar4   superscalar4

without Prefetch L1
0 0 0x01 -1 0 28592205443970 28592735589200
implementation   native         byteswap
scalar           3485493950     3145971812
superscalar      4925580275     4170573437
superscalar4     5476808168     4574764270
aarch64_neon     4804224295     4147109889
fastest          superscalar4   superscalar4
```